packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1827,9 +1827,59 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// actedUpdates() calls, and test on it for when to warn
+let actingUpdatesScopeDepth = 0;
+
+export function actedUpdates(callback: () => void | Promise<void>) {
+  let previousActingUpdatesScopeDepth;
   if (__DEV__) {
-    if (isRendering === false && isBatchingUpdates === false) {
+    previousActingUpdatesScopeDepth = actingUpdatesScopeDepth;
+    actingUpdatesScopeDepth++;
+  }
+
+  const result = batchedUpdates(callback);
+  if (result && result.then) {
+    // saving a few bytes without the typeof === 'function' check
+    return result.then(() => {
+      if (__DEV__) {
+        if (actingUpdatesScopeDepth - 1 > previousActingUpdatesScopeDepth) {
```

Comment:
An option to support some interleaving is to instead only do the check if you were the call that first acquired the semaphore, but I'm not 100% sure that is safe.

Reply:
I don’t think we want to support interleaving, do you have an example of when you’d want to interleaved test interactions?